### DISCUSSION TITLE
@QuarkusIntegrationTest Executing against a running application does not find Config/application.properties

### DISCUSSION THREAD

**Discussion by jochenr** [quarkusio/quarkus#33249] [Q&A]
Hello,

I'm struggling with my integration test that works fine, if I just run it with "mvn verify" and the test itself starts my Quarkus application.

Now I want to do what's decribed here:
[executing-against-a-running-application](https://quarkus.io/guides/getting-started-testing#executing-against-a-running-application)

Unfortunately starting my Integration-Test with parameters "-Dquarkus.http.test-host=someHost -Dquarkus.http.test-port=somePort" fails, when it reaches the code reading my application-properties
ConfigProvider.getConfig().getValue("someKey", String.class);

I tried to put a copy of my application.properties to "src/test/resources", but this didn't help.

How can I execute my QuarkusIntegrationTests against a running application so that it still uses the application.properties configuration?

I need this feature for testing Quarkus applications on different stagings/servers from a CI/CD Pipeline.

Thank you,
Jochen

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
Can you attach a sample application showing exactly what you are trying to do?

**Comment by bulldog98** [comment]
The problem might be that you need a space between setting different parameter via  `-D` in the invokation

**Reply by geoand** [reply]
That's definitely needed! Good catch!

**Comment by jochenr** [comment]
thank you for the quick response.
I just double checkd...unfortunately it was just a copy/paste fail, when putting it here;-(

In my test it was correct.

I hope I finde the time to do a little reproducer and put a link here

**Comment by jochenr** [comment]
here's the reproducer:

https://github.com/jochenr/quarkus-integration-test-example
```
mvn clean verify -DskipITs=false
```
...is working

then:

1st terminal window  
```
mvn quarkus:dev
```

2nd terminal window  
```
mvn verify -Dquarkus.http.test-host=localhost -Dquarkus.http.test-port=8080
```

...ist not working:
[ERROR] Errors:
[ERROR]   GreetingResourceIT.testReadingConfig:18 » NoSuchElement SRCFG00014: The config property my.config.prop is required but it could not be found in any config source

**Reply by geoand** [reply]
`ConfigProvider.getConfig()` won't work in integration tests as you expect it to

**Reply by jochenr** [reply]
Why does it work in "normal" (not against running application) integration-test mode then?
IMHO it should work or should not work consistently. The "quarkus.http.test-host" parameter should not effect the way the test works, doesn't it?

What is an elegant way of using the Quarkus config logic in a test then?
It's not good, if I have to duplicate part of the files application.properties, application-{profile}.properties, etc. in my integration tests.

**Reply by geoand** [reply]
I tried your sample and the tests work equally well whether the application is started or not

**Comment by jochenr** [comment]
now I'm confused

you don't get the
```
[ERROR] GreetingResourceIT.testReadingConfig:18 » NoSuchElement SRCFG00014: The config property my.config.prop is required but it could not be found in any config source
```
error when running 
```
mvn verify -Dquarkus.http.test-host=localhost -Dquarkus.http.test-port=8080
```
?

**Reply by geoand** [reply]
No

**Reply by jochenr** [reply]
sorry for asking....
and you really see the line
```
....INFO  [de.joc.qua.int.tes.exa.GreetingResourceIT] (main) myConfigProp:  My Value
```
in the output console?

I cannot imagine the difference between our environemnts that leads to the different behaviour;-(

**Reply by geoand** [reply]
```
[INFO] Running de.jochenr.quarkus.integration.test.example.GreetingResourceIT
2023-05-11 12:03:08,064 INFO  [de.joc.qua.int.tes.exa.GreetingResourceIT] (main) myConfigProp:  My Value

```

**Reply by jochenr** [reply]
I just repeated my test at home on my private computer to be sure that I didn't have some special configuration on my office computer that I haven't put to github....

.....exactly the same result here as I previously had on my work computer



```
mvn verify -Dquarkus.http.test-host=localhost -Dquarkus.http.test-port=8080
```
with mvn quarkus:dev running in second terminal window

result:
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running de.jochenr.quarkus.integration.test.example.GreetingResourceIT
[ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 5.389 s <<< FAILURE! - in de.jochenr.quarkus.integration.test.example.GreetingResourceIT
[ERROR] de.jochenr.quarkus.integration.test.example.GreetingResourceIT.testReadingConfig  Time elapsed: 0.009 s  <<< ERROR!
java.util.NoSuchElementException: SRCFG00014: The config property my.config.prop is required but it could not be found in any config source
        at io.smallrye.config.SmallRyeConfig.convertValue(SmallRyeConfig.java:298)
        at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:242)
        at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:167)
        at de.jochenr.quarkus.integration.test.example.GreetingResourceIT.testReadingConfig(GreetingResourceIT.java:18)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
```

why does 
```
ConfigProvider.getConfig()....
```
work on your computer in an QuarkusIntegrationTest with the failsafe plugin??

**Reply by geoand** [reply]
I think I tried the latest Quarkus instead of whatever version your sample was using

**Reply by jochenr** [reply]
latest means version 3.0.2 ?
I'll try this, too....

**Reply by geoand** [reply]
3.0.3

**Comment by jochenr** [comment]
Yes, with version 3.0.3.Final I get the same result as you do.
And the result I'd expect.

So I think I'll have to open a support case now to get it backportedm since I'm using (I have to use) Red Hat build of Quarkus 2.13.x ....

Or is my expectation wrong as you first stated this morning?

**Reply by geoand** [reply]
That was a feature we never advertised as something that would work.

However, you can certainly open a support case (and point to this discussion) and we can decide whether it makes sense to backport the change that makes the tests pass (if that is at all possible)

**Reply by jochenr** [reply]
thank you very much for your help👍

do you have another idea to make something as the Quarkus/Microprofile configuration work with "profiles" in integration-tests?
implementing it ourselves would be cumbersome...and actually the feature is already there;-)

**Reply by geoand** [reply]
You would likely build the configuration object manually - how many things you want to do depends on what you are trying to achieve - see https://github.com/quarkusio/quarkus/blob/5ab553a92cb6a3f4b89ee69781a329c81f033ed3/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java#L105
